---
title: "Naive T cell scRNA-seq import and analysis"
authors: "Dan Bunis, Rebecca Jaszczak"
date: "10/10/2023"
output:
  html_document:
    toc: true
    theme: united
---

[Original published analysis script](https://github.com/dtm2451/ProgressiveHematopoiesis/blob/master/scRNAseq_naiveTcells.Rmd) on which this is based. This script leans heavily on the Seurat [PBMC3K tutorial](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html) and the Seurat [Cell Cycle Vignette](https://satijalab.org/seurat/articles/cell_cycle_vignette.html) and a previous [sCRNA_Seq](https://github.com/UCSF-DSI/scRNA_seq) workshop Rebecca co-developed, with multiple sections directly copy and pasted.

```{r setup & libraries, include=FALSE}
#set some Rmd defaults
knitr::opts_chunk$set(echo = TRUE, fig.height = 6, fig.width = 7, warning = F)

#load in libraries
library(dittoSeq)
library(tidyverse)
library(Seurat)
```

# Read in raw, demultiplexed, downsampled data

Normally, you'll get FASTQ data off a sequencer, and run [the cellranger pipeline](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger) to link all the reads to their cells using the unique molecular identifiers (UMIs). Cellranger exports a `barcodes.tsv`, `matrix.mtx`, and `features.tsv`. Seurat provides the `Read10X()` command to initialize a Seurat object from these initial files. Rebecca has already done this, as well as downsampling, to make this object managable for a class. We'll read it in now!

```{r load in raw_data}
Tcells <- readRDS(file = './../raw_data/downsampled_Tcells_raw.rds')
Tcells
```

# Inspect data before filtering

Low-quality cells or empty droplets will often have very few genes. Cell doublets or multiplets may exhibit an aberrantly high gene count. (Note, these data were [previously multiplexed](https://github.com/statgen/popscle/wiki) and doublets have already been removed using `dittoSeq` by Rebecca via `tcell_downsample.R` script.) Similarly, the total number of molecules, or `nCount_RNA` detected within a cell (correlates strongly with unique genes, `nFeature_RNA`).

Low-quality / dying cells often exhibit extensive mitochondrial contamination as well. We use `PercentageFeatureSet()` to calculates the percentage of counts originating from a set of features.

We'll visualize the spread of our data using `dittoSeq` functions.

```{r metadata inspection}
#inspect metadata
head(Tcells@meta.data)

head(Tcells@assays$RNA@counts)
```


```{r qc assessment}
Tcells[["percent.mito"]] <- PercentageFeatureSet(object = Tcells,
                                                 pattern = "^MT-")

pre_filt_vlns <- multi_dittoPlot(object = Tcells,
                                 vars = c("nFeature_RNA", "nCount_RNA","percent.mito"),
                                 group.by = "orig.ident",
                                 plots = c("vlnplot","jitter"),
                                 jitter.width = 1.5,
                                 jitter.size = 0.25)
plot(pre_filt_vlns)

pre_filt_mito <- dittoScatterPlot(object = Tcells, 
                                  x.var = "nFeature_RNA",
                                  y.var = "percent.mito")
plot(pre_filt_mito)

pre_filt_count <- dittoScatterPlot(object = Tcells, 
                                   x.var = "nFeature_RNA",
                                   y.var = "nCount_RNA")
plot(pre_filt_count)
```

# Subset data and plots to visualize

Here we filter cells using the `subset()` function, to remove cells that have feature counts under 750, UMIs under 1500, and those that have more than 5% mitochondrial counts. Normally, one will filter cells with extremely high reads, but this has already been done.      

```{r subsetting}
Tcells <- subset(x = Tcells,
                 subset = nFeature_RNA > 750 & nCount_RNA > 1500 & percent.mito < 5)

post_filt_count <- dittoScatterPlot(object = Tcells,
                                    x.var = "nFeature_RNA",
                                    y.var = "nCount_RNA")

post_filt_mito <- dittoScatterPlot(object = Tcells, 
                                   x.var = "nFeature_RNA",
                                   y.var = "percent.mito")
```

`dittoScatterPlot` function is used here to visualize feature-feature (gene-gene) relationships, but can be used for anything calculated by the object, e.g. columns in object metadata, PC scores etc.

```{r visualize filtering}
pre_filt_count + post_filt_count
pre_filt_mito + post_filt_mito

#add a demo of grid of 4 plots
(pre_filt_count + post_filt_count) /
      (pre_filt_mito + post_filt_mito)
```

# Add some metadata

Next, we'll add some metadata slots to help us parse the data. 

Seurat metadata is stored in the `object@meta.data` slots. It's just a normal data frame, and you can use tidyverse to manipulate it! We're going to add slots for `Age` (whether the sample was `fetal`, `cord`, or `adult`), T cell type (`CD4` or `CD8`), and a `Tage` variable combining both. T cell type is dependent on both sample-assignment and 10X-lane, and copied from Dan's code. We'll also add a cell cycle meta column, to regress against cell cycle signatures which can dominate gene expression.

## Age

Here, we create a new column `Age` in the `@meta.data` which is assigned `fetal` if the `Sample` name starts with `FS`. We select samples with the `FS` designation using a function called `grepl`. We apply the same principle to `cord` samples starting with `UCB`, and `adult` samples starting with `APB`.

```{r inspect all metadata colnames}
#inspect unique metadata in "SAMPLE"
unique(Tcells@meta.data$Sample)
colnames(Tcells@meta.data)
```


```{r add age to meta}
Tcells@meta.data <- Tcells@meta.data %>%
  mutate(Age = case_when(grepl("APB", Sample) ~ 'adult',
                         grepl('FS', Sample) ~ 'fetal',
                         grepl('UCB', Sample) ~ 'cord'))

unique(Tcells@meta.data$Age)
```

## T Cell Type

Here, we assign a cell type based on Dan's information from how he generated these libraries. We select `Sample` values from a vector using `%in%` and then further select those samples based on which sequencing `Lane` they were run on.

```{r add t cell type to meta}
# Add Tcelltype
samples.4 <- c("FS3", "FS4", "FS5", "UCB1", "UCB2", "UCB5", "APB1", "APB2", "APB4", "APB5")
samples.8 <- c("FS1", "FS5", "UCB2", "UCB3", "UCB4", "UCB5", "APB2", "APB3", "APB5")
samples.48.4 <- c("FS1", "FS2", "UCB4", "UCB3", "APB3") 
samples.48.8 <- c("FS3", "UCB1", "APB1", "APB4")
samples.48 <- c(samples.48.4,samples.48.8)

Tcells@meta.data <- Tcells@meta.data %>%
  mutate(Tcelltype = case_when(Sample %in% samples.4 & Lane == "CD4" ~ "CD4",
                               Sample %in% samples.48.4 & Lane == "CD4-8" ~ "CD4",
                               Sample %in% samples.8 & Lane == "CD8" ~ "CD8",
                               Sample %in% samples.48.8 & Lane == "CD4-8" ~ "CD8"))

unique(Tcells@meta.data$Tcelltype)
```

## Age and T Cell type combined

We can `unite` the `Tcelltype` and `Age` variables in order to generate a combined meta column for future plots. Make sure to add the `remove = FALSE` flag, or the original columns are deleted from the `@meta.data`

```{r add t cell age to meta}
#Add a Tage metadata so that I can plot things with the ages and CD4s vs CD8s separately:
Tcells@meta.data = Tcells@meta.data %>%
      unite('Tage', Age:Tcelltype, remove = FALSE)

#write unique query function
unique(Tcells@meta.data$Tage)
```

## Cell Cycle

We can use the `ridgeplot` specification from `dittoSeq` to visualize some genes and show their expression.

```{r add cell cycle scoring to meta}
Tcells <- CellCycleScoring(object = Tcells,
                           s.features = cc.genes$s.genes,
                           g2m.features = cc.genes$g2m.genes,
                           set.ident = TRUE)

#save the idents to a new CellCycle meta column
Tcells@meta.data$CellCycle <- Idents(Tcells)

# Visualize the distribution of cell cycle markers across
genes_to_plot = c('PCNA', 'NASP', 'HMGB2', 'SMC4', 'TUBB4B')

multi_dittoPlot(object = Tcells, 
          var = genes_to_plot,
          plots = c("ridgeplot"),
          group.by = 'CellCycle')
```

# Visualize after filtering

Here, we use the `multi_dittoPlot` function to visualize violin plots with dots for each cell, and a `dittoScatterPlot` to display these data in another way.

Finally, we can use the `table` function to see how many individual cells from each sample are in each category.

```{r filtered plots}
multi_dittoPlot(object = Tcells,
                vars = c("nFeature_RNA", "nCount_RNA","percent.mito"),
                group.by = "orig.ident",
                plots = c("vlnplot","jitter"),
                jitter.width = 1.5,
                jitter.size = 0.25)

# Better examined with:
dittoScatterPlot(Tcells,
                 "nFeature_RNA", "nCount_RNA", color.var = "percent.mito",
                 max = 20) +
    geom_hline(yintercept = 1500) +
    geom_vline(xintercept = 750)

# table(Tcells$Sample,Tcells$Lane,Tcells$Tcelltype)
table(Tcells@meta.data$Tage)
```

# Log Transform Data

After removing unwanted cells from the dataset, the next step is to normalize the data. By default, we employ a global-scaling normalization method “LogNormalize” that normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result. Normalized values are stored in `object[["RNA"]]@data`.


```{r normalize data}
Tcells <- NormalizeData(
  object = Tcells,
  normalization.method = "LogNormalize",
  scale.factor = 10000)

#Tcells <- NormalizeData(Tcells)
```
Note: For clarity, in this previous line of code (and in future commands), we provide the default values for certain parameters in the function call. Look at the help file by running `?NormalizeData` in the console. This will bring up documentation for the function, with the default values and what they mean. For example, the "LogNormalize" argument means that feature "counts for each cell are divided by the total counts for that cell and multiplied by the scale.factor. This is then natural-log transformed using log1p.". You can try different methods.

```{r display count data}
Tcells@assays$RNA@counts

```

```{r display normalized data}
Tcells@assays$RNA@data

```

# Pick variable genes that will be used for PCA & tSNE

We next calculate a subset of features that exhibit high cell-to-cell variation in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others). The Seurat developers and [others](https://www.nature.com/articles/nmeth.2645) have found that focusing on these genes in downstream analysis helps to highlight biological signal in single-cell datasets.

By default, we return 2,000 features per dataset and we use the "vst" method which fits a local polynomial regression line (see ?FindVariableFeatures for other methods). These will be used in downstream analysis, like PCA.  


```{r find variable features}
Tcells <- FindVariableFeatures(object = Tcells,
                               selection.method = "vst",
                               nfeatures = 2000)

# plot variable features with and without labels
LabelPoints(plot = VariableFeaturePlot(Tcells), 
            points = head(VariableFeatures(Tcells), 10),
            repel = TRUE)
```

Next, we apply a linear transformation ("scaling") that is a standard pre-processing step prior to dimensional reduction techniques like PCA. The `ScaleData()` function:

  - Shifts the expression of each gene, so that the mean expression across cells is 0
  - Scales the expression of each gene, so that the variance across cells is 1
  - This step gives equal weight in downstream analyses, so that highly-expressed genes do not dominate
  - The results of this are stored in `object[["RNA"]]@scale.data`

We can also take care of unwanted sources of variation in this step. Some things to think about regressing include:

  - cell-cell variation in gene expression driven by batch (for example, two 10X runs from different days)
  - cell cycle  
  - number of detected molecules
  - mitochondrial gene expression  

```{r scale and regress data}
#Scale without regressing out any variables
Tcells <- ScaleData(
  object = Tcells,
  vars.to.regress = c("CellCycle", "nCount_RNA", "percent.mito"))
```
# Saving and reading output     

You can save the object using `saveRDS()` at this point so that it can easily be loaded back in without having to rerun the computationally intensive steps performed above. An example of how to read the data back in is also provided, although commented out.

```{r save point!}
#create a "processed_data" dir
dir.create('./../processed_data')
saveRDS(Tcells, file = "../processed_data/Tcells_analyzed.rds")
# Tcells <- readRDS("../processed_data/Tcells_analyzed.rds"))
```

# Run PCA & UMAP

Next we perform Principle Component Analysis (PCA) on the scaled data. By default, only the previously determined variable features are used as input, but can be defined using `features` argument if you wish to choose a different subset. Remember that in the previous notebook we used the `VariableFeatures()` function and we got a vector of the 2000 most variable genes.    
```{r PCA calculation}
#Run PCA
Tcells <- RunPCA(
  object = Tcells,
  verbose = TRUE,
  npcs = 50)
```

```{r}
Idents(Tcells) = 'Age'

DimPlot(Tcells,
        dims = c(1,2),
        reduction = 'pca')
```

To overcome the extensive technical noise in any single feature for scRNA-seq data, Seurat clusters cells based on their PCA scores, with each PC essentially representing a ‘metafeature’ that combines information across a correlated feature (gene) set. The top principal components therefore represent a robust compression of the dataset. However, how many componenets should we choose to include? 10? 20? 100?

We will use PCheatmaps, PCElbowPlot, and JackStrawPlot to pick which PCs to carry forward into tSNE and clustering.

A commonly used heuristic method generates an ‘Elbow plot’: a ranking of principle components based on the percentage of variance explained by each one (`ElbowPlot` function). In this example, we can observe an ‘elbow’ around PC9-10, suggesting that the majority of true signal is captured in the first 10 PCs. This is a classic visualization for PCs.


```{r PCA selection}
PCHeatmap(
   object = Tcells,
   dims = 1:12, 
   cells = 2000, 
   balanced = TRUE
)

ElbowPlot(object = Tcells)

# this code took 4 mins to run on a machine with 32gb ram
# recommend running on your own time
# Tcells <- JackStraw(object = Tcells, num.replicate = 100, dims = 20, verbose = TRUE)
# Tcells <- ScoreJackStraw(Tcells, dims = 1:20)
# JackStrawPlot(Tcells, dims = 1:20)
```
The large change in p value magnitude from 9 to 10, and 9 pcs being around an inflection point in the elbow plot went into our picking PCs1-9 for clustering and umap.

Seurat v3 constructs a KNN graph based on the euclidean distance in PCA space, and refines the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). This step is performed using the `FindNeighbors` function, and takes as input the previously defined dimensionality of the dataset (first 9 PCs).

To cluster the cells, Seurat v3 next applies modularity optimization techniques such as the `Louvain algorithm` (default) to iteratively group cells together. The `FindClusters` function implements this procedure.

We find that setting this parameter between `0.4-1.2` typically returns good results for single-cell datasets of around 3K cells. Optimal resolution often increases for larger datasets. 

```{r Clustering}
Tcells <- FindNeighbors(object = Tcells,
                        reduction = "pca",
                        k.param = 20,
                        dims = 1:9)
Tcells <- FindClusters(object = Tcells,
                       modularity.fxn = 1,
                       algorithm = 1,
                       resolution = 0.1)
```

Seurat offers several non-linear dimensional reduction techniques (including `UMAP`, the most commonly used) to visualize and explore datasets. The goal of these algorithms is to place similar cells together in low-dimensional space. We suggest using the same PCs as input to the clustering analysis.  

# Run UMAP
```{r run UMAP}
Tcells <- RunUMAP(
  object = Tcells,
  reduction = "pca",
  dims = 1:9,
  seed.use = 1337)
```

```{r plot UMAP}
#Colored with CD4s lighter and CD8s darker 
dittoDimPlot(Tcells, "Tage", size = 1, reduction.use = "umap",
          colors = c(1:3,9:11), main = "T cells Lineage and Stage",
          rename.var.groups = c("Adult-CD4", "Fetal-CD4", "UCB-CD4",
                            "Adult-CD8", "Fetal-CD8", "UCB-CD8"))
```